### PR TITLE
Added snippet to show how to override conversation ID header value.

### PR DESCRIPTION
**Pull request by ramonsmits**

### CODE REVIEW

**Comment by ramonsmits** [comment]
@andreasohlund I'm trying to override the value, but it doesn't work.

> Cannot set the NServiceBus.ConversationId header to 'XXXXX' as it cannot override the incoming header value ('2f4076a0-d8de-4297-9d18-a830015dd42a').

Why not??

**Comment by andreasohlund** [comment]
We concluded that overriding should only be happening when sending outside
messagehandlers. Do you have a use case for overriding on
handlers?(essentially ”breaking” the message flow chain)

**Comment by andreasohlund** [comment]
@timbussmann I remember that you had some concerns that we're being to strict?

File: Snippets/Core/Core_7/Headers/OverrideConversationId.cs [hunk 1]
```diff
@@ -0,0 +1,17 @@
+﻿using System.Threading.Tasks;
+using Core7.Handlers;
+using NServiceBus;
+
+public class ConversationIdHeaderOverride
```
Comment (andreasohlund) [inline Snippets/Core/Core_7/Headers/OverrideConversationId.cs]: @ramonsmits I'm about to add doco for https://github.com/Particular/NServiceBus/blob/74a1c4bddfc5167deeb8fcd010fe0644f25823fa/src/NServiceBus.AcceptanceTests/Core/Causation/When_overriding_conversation_id_generation.cs#L53 

I wonder if we should make that the only documented option for v7? (or is there still a usecase for "poluting" the business code with setting the header)

File: nservicebus/messaging/headers.md [hunk 2]
```diff
@@ -89,6 +89,11 @@ Identifier of the conversation that this message is part of. It enables the trac
 
 The first message that is sent in a new flow is automatically assigned a unique `Conversation Id` that is then propagated to all the messages that are subsequently sent, thus forming a _conversation_. Each message that is sent within a conversation also has a `RelatedTo` value that identifies the originating message that caused it to be sent. 
 
+There are scenarios where you want to override the Conversation Id because having a single huge conversation doesn't make sense. This can be done by overriding the header with a new value:
```
Comment (andreasohlund) [inline nservicebus/messaging/headers.md]: I'd say its not about a single huge one but rather having a business related id instead to make things like `PlaceOrder` and `CancelOrder` show up in the same conversation (and display nicely in SI) by using the OrderId as the conversation id?

I was planning to pull this more philosophical part out to a dedicated page as part of https://github.com/Particular/docs.particular.net/issues/3311

So perhaps we can shorten this to "Overriding the ConversationId is done setting the conversation id header value when sending a message"

Then I can deal with the full explanation as part of #3311?

**Comment by timbussmann** [comment]
@andreasohlund yeah I was wondering about the risk of allows allowing overwriting the conversationid. But so far I think we haven't heard from users having issues with the current behavior.

File: Snippets/Core/Core_7/Headers/OverrideConversationId.cs (hunk 1 above)
Comment (ramonsmits) [reply Snippets/Core/Core_7/Headers/OverrideConversationId.cs]: To be honest, I think the business code is the perfect place to indicate what the conversation ID is. I personally would want this magic per message to be hidden in a mutator or pipeline unless it can be really applied generically.

Comment (ramonsmits) [reply Snippets/Core/Core_7/Headers/OverrideConversationId.cs]: For example, you could use a convention based on message ID, take the first part of a message ID to identify the conversation:

File: nservicebus/messaging/headers.md (hunk 2 above)
Comment (DavidBoike) [reply nservicebus/messaging/headers.md]: It can be about a single huge one. Picture a long-running process like a scheduler (or implemented as a "We finished one, send a message to do another") that periodically spawns "child conversations" where currently the entire conversation would span multiple months and thousands of messages.

Comment (ramonsmits) [reply nservicebus/messaging/headers.md]: THat is exactly what I meant with huge, its never ending saga with timeouts.

**Comment by ramonsmits** [comment]
@timbussmann We did, in the past I processed a case where this was an issue because in ServiceInsight it was impossible to look at the conversation.

**Comment by ramonsmits** [comment]
@andreasohlund updated the current document but still would like to see the ability to override the behavior. The only way to do that would I think manipulate the incoming headers someway or to use IMessageSession within a receive operation which both really shouldn't be needed.

**Comment by timbussmann** [comment]
> We did, in the past I processed a case where this was an issue because in ServiceInsight it was impossible to look at the conversation.

I think we had a v6 issue where it wasn't possible to set a custom id as it was always overriden by the pipeline at a later point.

File: Snippets/Core/Core_3/Headers/OverrideConversationId.cs [hunk 3]
```diff
@@ -0,0 +1,12 @@
+﻿using NServiceBus;
```
Comment (timbussmann) [inline Snippets/Core/Core_3/Headers/OverrideConversationId.cs]: why adding v3 snippets?

File: Snippets/Core/Core_6/Headers/OverrideConversationId.cs [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+using NServiceBus;
+
+public class ConversationIdHeaderOverride
+{
+    public async Task Handle()
+    {
+        IMessageHandlerContext context = null;
+        #region override-conversation-id
+        var sendOptions = new SendOptions();
+        sendOptions.SetHeader(Headers.ConversationId, "MyCustomConversationId/" + System.Guid.NewGuid());
```
Comment (timbussmann) [inline Snippets/Core/Core_6/Headers/OverrideConversationId.cs]: I think there is no scenario (except 3rd party integration) where an incoming message doesn't have a conversation id present. Given that this will fail, it doesn't seem to be a valid snippet to set a conversation id in a handler.

File: Snippets/Core/Core_3/Headers/OverrideConversationId.cs (hunk 3 above)
Comment (ramonsmits) [reply Snippets/Core/Core_3/Headers/OverrideConversationId.cs]: Because it wasn't any effort to add.

File: Snippets/Core/Core_6/Headers/OverrideConversationId.cs (hunk 4 above)
Comment (ramonsmits) [reply Snippets/Core/Core_6/Headers/OverrideConversationId.cs]: @timbussmann You are right, this was still under the assumption that is would work. I'll update the interface.

**Comment by andreasohlund** [comment]
>but still would like to see the ability to override the behavior. The only way to do that would I think manipulate the incoming headers someway or to use IMessageSession within a receive operation which both really shouldn't be needed.

Agreed, can you raise a core issue to track this?

**Comment by andreasohlund** [comment]
>pipeline unless it can be really applied generically.

That was the intent with guiding the users towards a custom attribute/interface/etc , see https://github.com/Particular/docs.particular.net/issues/3311

Ie to mark their business message in some way, but have the actual extraction be generic to not clutter the code that emits the message. Does that make sense?

File: nservicebus/messaging/headers.md [hunk 5]
```diff
@@ -89,6 +89,16 @@ Identifier of the conversation that this message is part of. It enables the trac
 
 The first message that is sent in a new flow is automatically assigned a unique `Conversation Id` that is then propagated to all the messages that are subsequently sent, thus forming a _conversation_. Each message that is sent within a conversation also has a `RelatedTo` value that identifies the originating message that caused it to be sent. 
 
+There are scenarios where you want to manually set the Conversation Id, for example, a CancelOrder needs to become part of an existing order conversation. This can be done by overriding the header with a custom value:
+
+snippet: override-conversation-id
+
+It is not possible to override an existing Conversation Id, trying to do this results in the following error:
```
Comment (andreasohlund) [inline nservicebus/messaging/headers.md]: Should this be a `NOTE:` ?

Perhaps also add that messages emitted from handlers will get the conversation id of the incoming message so its not really possible to override conversation ids from message handlers?

Comment (ramonsmits) [reply nservicebus/messaging/headers.md]: @andreasohlund The exception that follows it is part of it. I don't think that can be combined.

Comment (andreasohlund) [reply nservicebus/messaging/headers.md]: ok

**Comment by ramonsmits** [comment]
@andreasohlund raised https://github.com/Particular/NServiceBus/issues/5088

File: nservicebus/messaging/headers.md (hunk 5 above)
Comment (ramonsmits) [reply nservicebus/messaging/headers.md]: @andreasohlund the section already has the following:

**Comment by weralabaj** [comment]
@ramonsmits it's a bit hard to follow what's going on in this issue, please, once you resolve all the "content" issues and it's ready for docs maintainers review, request it explicitly to give us a signal, ok?

**Comment by andreasohlund** [comment]
Build fails with 

>System.Exception : Some redundant snippets were detected. override-conversation-id (>= 4.0.0 && < 5.0.0) override-conversation-id (>= 3.0.0 && < 4.0.0)
   at RedundantSnippets.ThrowForRedundantSnippets(IReadOnlyList`1 availableSnippets, IReadOnlyList`1 usedSnippetKeys) in C:\BuildAgent\work\3ba50a7be36aeb6b\src\ParticularDocsCore\Sample\RedundantSnippets.cs:line 15
   at ArticleSnippetsReader.Finish() in C:\BuildAgent\work\3ba50a7be36aeb6b\src\ParticularDocsCore\ArticleSnippetsReader.cs:line 33
   at SiteRunner.RenderSite(String sourceDirectory, String targetDirectory, String lastShaPrefix) in C:\BuildAgent\work\3ba50a7be36aeb6b\src\ParticularDocsCore\SiteRunner.cs:line 37
   at ContentVerificationTest.RunFullSiteGeneration() in C:\BuildAgent\work\3ba50a7be36aeb6b\src\DocsWeb.Tests\ContentVerificationTest.cs:line 22

**Comment by weralabaj** [comment]
the page has versions header saying it only covers version 5 and above so either we need to change that or remove the snippets

**Comment by weralabaj** [comment]
rebased
